core/hotspot/cache/wtinylfu/slru.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 1999-2020 Alibaba Group Holding Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wtinylfu
+
+import "container/list"
+
+type listType uint8
+
+const (
+	admissionWindow listType = iota
+	probationSegment
+	protectedSegment
+)
+
+const protectedRatio = 0.8
+
+type slruItem struct {
+	listType listType
+	key      interface{}
+	value    interface{}
+	keyHash  uint64
+}
+
+// slru is a segmented LRU.
+type slru struct {
+	data                       map[interface{}]*list.Element
+	probationCap, protectedCap int
+	probationLs, protectedLs   *list.List
+}
+
+func newSLRU(cap int, data map[interface{}]*list.Element) *slru {
+	protectedCap := int(float64(cap) * protectedRatio)
+	probationCap := cap - protectedCap
+	return &slru{
+		data:         data,
+		probationCap: probationCap,
+		probationLs:  list.New(),
+		protectedCap: protectedCap,
+		protectedLs:  list.New(),
+	}
+}
+
+// access access a value from the cache
+func (slru *slru) access(v *list.Element) {
+	item := v.Value.(*slruItem)
+	if item.listType == protectedSegment {
+		slru.protectedLs.MoveToFront(v)
+		return
+	}
+	if slru.protectedLs.Len() < slru.protectedCap {
+		slru.probationLs.Remove(v)
+		item.listType = protectedSegment
+		slru.data[item.key] = slru.protectedLs.PushFront(item)
+		return
+	}
+	back := slru.protectedLs.Back()
+	backItem := back.Value.(*slruItem)
+
+	// swap the two item
+	*backItem, *item = *item, *backItem
+	backItem.listType = protectedSegment
+	item.listType = probationSegment
+	slru.data[item.key] = v
+	slru.data[backItem.key] = back
+
+	// move the elements to the front of their lists
+	slru.probationLs.MoveToFront(v)
+	slru.protectedLs.MoveToFront(back)
+}
+
+// add set a value in the cache
+func (slru *slru) add(newItem slruItem) {
+	newItem.listType = probationSegment
+	if slru.probationLs.Len() < slru.probationCap || slru.Len() < slru.probationCap+slru.protectedCap {
```

Comment:
How to understand the condition `|| slru.Len() < slru.probationCap+slru.protectedCap` ?

Reply:
This logic is because if there is only add but no get, probation lru cannot be promoted to protected lru, at this time all elements are in probation